Dockerfile

```diff
@@ -1,36 +1,71 @@
-FROM archlinux:base-devel
+FROM archlinux:base-devel AS planner
+RUN pacman -Syy --noconfirm
 
-ENV MPLCONFIGDIR="/home/app_user/"
-ARG DEPENDENCIES="mkvtoolnix curl llvm clang"
+# Install all dependencies (except for rav1e)
+RUN pacman -S --noconfirm rsync rust clang nasm git aom ffmpeg vapoursynth ffms2 libvpx mkvtoolnix-cli svt-av1 vapoursynth-plugin-lsmashsource vmaf
+
+WORKDIR /tmp/Av1an
+RUN cargo install cargo-chef 
+COPY . .
+RUN cargo chef prepare  --recipe-path recipe.json
+
+
+
+
+FROM archlinux:base-devel AS cacher
+RUN pacman -Syy --noconfirm
+
+# Install all dependencies (except for rav1e)
+RUN pacman -S --noconfirm rsync rust clang nasm git aom ffmpeg vapoursynth ffms2 libvpx mkvtoolnix-cli svt-av1 vapoursynth-plugin-lsmashsource vmaf
+
+WORKDIR /tmp/Av1an
+RUN cargo install cargo-chef
+COPY --from=planner /tmp/Av1an/recipe.json recipe.json
+RUN cargo chef cook --release --recipe-path recipe.json
+
+
+
+
+FROM archlinux:base-devel AS build
 
 RUN pacman -Syy --noconfirm
 
-# Install make dependencies
-RUN pacman -S --noconfirm rust clang nasm git
+# Install all dependencies (except for rav1e)
+RUN pacman -S --noconfirm rsync rust clang nasm git aom ffmpeg vapoursynth ffms2 libvpx mkvtoolnix-cli svt-av1 vapoursynth-plugin-lsmashsource vmaf
+
+# Compile rav1e from git, as archlinux is still on rav1e 0.4
+RUN git clone https://github.com/xiph/rav1e /tmp/rav1e
+WORKDIR /tmp/rav1e
+RUN cargo build --release && \
+    strip ./target/release/rav1e 
+RUN mv ./target/release/rav1e /usr/local/bin
+
+# Build av1an
+COPY . /tmp/Av1an
+
+# Copy over the cached dependencies
+COPY --from=cacher /tmp/Av1an/target /tmp/Av1an/target
+
+WORKDIR /tmp/Av1an
+RUN cargo build --release
+RUN mv ./target/release/av1an /usr/local/bin
+
+
+
+FROM archlinux:base-devel AS runtime
+
+ENV MPLCONFIGDIR="/home/app_user/"
+
+RUN pacman -Syy --noconfirm
 
 # Install all optional dependencies (except for rav1e)
 RUN pacman -S --noconfirm aom ffmpeg vapoursynth ffms2 libvpx mkvtoolnix-cli svt-av1 vapoursynth-plugin-lsmashsource vmaf
 
-# Compile rav1e from git, as archlinux is still on rav1e 0.4
-RUN git clone https://github.com/xiph/rav1e && \
-    cd rav1e && \
-    cargo build --release && \
-    strip ./target/release/rav1e && \
-    mv ./target/release/rav1e /usr/local/bin && \
-    cd .. && rm -rf ./rav1e
+COPY --from=build /usr/local/bin/rav1e /usr/local/bin/rav1e
+COPY --from=build /usr/local/bin/av1an /usr/local/bin/av1an
 
 # Create user
 RUN useradd -ms /bin/bash app_user
-
-# Copy av1an and build av1an
-COPY --chown=app_user . /Av1an
-WORKDIR /Av1an
-RUN cargo build --release && \
-    mv ./target/release/av1an /usr/local/bin && \
-    cd .. && rm -rf ./Av1an
-
-# Remove build dependencies
-RUN pacman -R --noconfirm rust clang nasm git
```

Comment:
Is it possible to add back this line? The build dependencies are not needed at runtime, so the container size could be further reduced.

Reply:
Removing anything afterwards doesnt affect the size of the docker container as those still exist in the layers prior due to how docker works. Those build dependencies actually do not exist in the final runtime due to them only existing in the build stage and is the benefit of using multistaging.